### DISCUSSION TITLE
Strange Oval when combining and distorting images

### DISCUSSION THREAD

**Discussion by mathiasgrimm** [ImageMagick/ImageMagick#6337] [Help]
### ImageMagick version

6.x.x or 7.x.x

### Operating system, version and so on

MacOs or Linux

### Description

Hi,
Sometimes when using the distort and combine I get some weird ovals over the image.

In this example, I have two images:frame.jpg that's the frame, and another image $imagePath.
from the $imagePath I create another image to place at the left side of the frame which I connect the points with the $imagePath.

Image that will be place on the frame
![black-2](https://github.com/ImageMagick/ImageMagick/assets/450069/9702cb04-833a-413e-8d95-913cd36b6d6e)

<img width="685" alt="image" src="https://github.com/ImageMagick/ImageMagick/assets/450069/5fc317a8-cc5d-484f-a91d-bdf5ee5c3265">

Result:
<img width="972" alt="image" src="https://github.com/ImageMagick/ImageMagick/assets/450069/d6e8b970-46af-4bae-b36d-ce9076d8ff15">

Sometimes playing a bit with the connecting pixels (+1, -1) fixes it but so far I haven't been able to resolve it in a deterministic way.

in this example, if I comment out the line that has the `TODO Incomplete` it works, in the sense that the oval disappears however the resulting image is will have a small positioning issue.

Following is the main part of a php code that does it:
```php
<?php
    $imagePath = 'some-image.jpg';
    $framePath = __DIR__ . '/frame.jpg';
    $compositeType = Imagick::COMPOSITE_MULTIPLY;

    $w = 1250;
    $h = 1250;

    $frame = new Imagick($framePath);

    $imageOriginal = new Imagick($imagePath);

    $imageOriginal->resizeimage(
        $w,
        $h,
        Imagick::FILTER_LANCZOS,
        true,
    );

    $front = $imageOriginal->clone();

    $front->resizeimage(
        $w,
        $h,
        Imagick::FILTER_LANCZOS,
        true,
    );

    $leftSide = $imageOriginal->clone();
    $leftSide->cropImage(
        1,
        $front->getImageHeight(),
        0,
        0,
    );

    $leftSide->resizeimage(
        $w,
        $h,
        Imagick::FILTER_LANCZOS,
        true,
    );

    $topX = 125;
    $topY = 247;

    $bottomX = 182;
    $bottomY = 998;

    $controlPointsLeft = [
        0, 0, 123, 250, // top left
        $w, 0, $topX, $topY, // top right - MATCH
        $w, $h, $bottomX, $bottomY, // bottom right - MATCH
        0, $h, 174 , 992, // bottom left  // TODO Incomplete
    ];
    $controlPointsFront = [
        0, 0, $topX - 1, $topY, // top left - MATCH
        $w, 0, 1049, 251, // top right
        $w, $h, 1086, 851, // bottom right
        0, $h, $bottomX - 1, $bottomY, // bottom left - MATCH
    ];

    $front->distortImage(Imagick::DISTORTION_BILINEAR, $controlPointsFront, false);
    $leftSide->distortImage(Imagick::DISTORTION_BILINEAR, $controlPointsLeft, false);

    $front->compositeImage($leftSide, Imagick::COMPOSITE_DEFAULT, 0, 0);
    $frame->compositeImage($front, $compositeType, 0, 0);
```

**Comment by fmw42** [comment]
Change the virtual pixel to black and see if that oval turns black?  See the examples at https://www.php.net/manual/en/imagick.distortimage.php for how to change the virtual pixel.

**Reply by mathiasgrimm** [reply]
Hey, thanks!

I think we might be getting somewhere.

That's the result:
<img width="855" alt="image" src="https://github.com/ImageMagick/ImageMagick/assets/450069/07567429-1527-46ba-9c8a-fe43f62e2437">


Changed to be like this:
<img width="749" alt="image" src="https://github.com/ImageMagick/ImageMagick/assets/450069/9c9106f3-f995-48a7-9f68-9bb13067e940">


Even when I uncomment the two commented lines the result is the same as the above

**Reply by fmw42** [reply]
That is not the result that I expected.  I thought the oval would turn black and the rest of the result would look the same.

**Reply by fmw42** [reply]
You should use a perspective distortion to warp the black image onto the frame, rather than a bilinear warp.

**Reply by mathiasgrimm** [reply]
The Perspective doesn't always align correctly but I also get this when using it.
<img width="397" alt="image" src="https://github.com/ImageMagick/ImageMagick/assets/450069/a189eb66-eb77-4e23-bb7b-20ca05ae21c1">

**Comment by fmw42** [comment]
I am not sure what you are doing with two distortions.  But if you want to put the black image onto the frame image, do the following in command line Unix syntax.  My control points look different from yours!

```
magick frame.png +repage \
\( black.jpg -virtual-pixel none +distort perspective \
"0,0 143,273  %w,0 1198,278  %w,%h 1239,962  0,%h 207,1129" \) \
-layers merge +repage \
frame_black.jpg
```

![frame_black](https://github.com/ImageMagick/ImageMagick/assets/13458218/ee9127bc-f7c8-47ab-9fb9-6c3ccafa69d6)

**Reply by mathiasgrimm** [reply]
the side (spine) is a third image which is a clone of the black one, but with 1px of width.
i resize both the black (front) and the side with the size of the frame so that the pixel reference is constant as in, 200,200 would always be the same place acros the 3 images. In this case a black image doesn't show the difference and it works for the code you pasted but in reality it won't be a black one and it will lose the 3d aspect if we don't have a separate image for the side.

This works 99% of the time, but in some cases i get these weird ovals. It really depends

**Reply by mathiasgrimm** [reply]
https://legacy.imagemagick.org/Usage/distorts/#box3d

**Reply by fmw42** [reply]
You likely need at least 2 pixels wide image, not just one pixel

**Reply by mathiasgrimm** [reply]
Im resizing a 1px image into 1250x1250

```
$leftSide = $imageOriginal->clone();
        $leftSide->cropImage(
            1,
            $front->getImageHeight(),
            0,
            0,
        );

        $leftSide->resizeimage(
            $w,
            $h,
            Imagick::FILTER_LANCZOS,
            true,
        );
```

**Reply by fmw42** [reply]
Sorry, I do not use Imagick. I am trying to make the equivalent in command line, but I do not understand your control points. They are not the same as what I used above.  Also can you provide an example that works, with all your images, so that I can see how the result should look. I still do not fully understand what you are doing.  You want one image inside the frame and the other on the left side of the frame. Is that correct?  How should the left side image show on your result?

**Reply by mathiasgrimm** [reply]
another example:

image to be placed on the frame:
![testphp2](https://github.com/ImageMagick/ImageMagick/assets/450069/700052fb-7e5e-4ff1-a3df-21e370763341)


frame:
<img width="680" alt="image" src="https://github.com/ImageMagick/ImageMagick/assets/450069/d4c1d651-fcfd-4bc8-b592-1f9fe940042e">

result (correct):
<img width="992" alt="image" src="https://github.com/ImageMagick/ImageMagick/assets/450069/ad3dd85a-4aa2-45fb-96df-6655453d3c9e">

**Reply by mathiasgrimm** [reply]
please observe carefully the left side on the frame as it lines up and gives that 3d look

**Reply by fmw42** [reply]
What image is used for the left side?  Is it a crop of the first image or something else?   

Why did neither of your control point sets seem to match what I used in the command line? 

Please provide code (control point arguments) that match your example images, so we can try to reproduce your work.

**Comment by fmw42** [comment]
Note that 3D box used +distort (perspective) and -layers merge.  You are using -distort (bestfit false) bilinear and normal composite rather than mergeImageLayers().  That may be the cause of your issues.
